File: Roadmap/10 - EXCEPCIONES/javascript/cesar-ch.js
/*
    #10 EXCEPCIONES
*/

try {
    const x = 10 / 0
    console.log(`El resultado de la division es: ${x}`)
} catch (error) {
    console.log(`Error: ${error.message}`)
} finally {
    console.log('Ejecución finalizada')
}

/*
    DIFICULTAD EXTRA 
*/

class ErrorPersonalizado extends Error {
    constructor(message) {
        super(message);
        this.name = 'ErroPersonalizado';
    }
}

function procesarParametros(parametro1, parametro2) {
    if (typeof parametro1 !== "number" || typeof parametro2 !== "number") {
        throw new ErrorPersonalizado('Los parametros deben ser numericos')
    }
    if (parametro2 === 0) {
        throw new Error('No se puede dividir por cero')
    }
    return parametro1 / parametro2
}

try {
    const resultado = procesarParametros("10", 0)
    console.log(`El resultado de la division es: ${resultado}`)
} catch (error) {
    if (error instanceof ErrorPersonalizado) {
        console.log(`Error: ${error.message}`)
    } else {
        console.log(`Error: ${error.message}`)
    }
} finally {
    console.log('Ejecución finalizada')
}

File: Roadmap/10 - EXCEPCIONES/python/cesar-ch.py
"""
    #10 EXCEPCIONES
"""

try:
    resultado = 10 / 0
except Exception as e:
    print(f"Ocurrió un error: {e}")
else:
    print(f"El resultado es: {resultado}")
finally:
    print("La ejecución ha finalizado.")


lista = ["a", "b", "c"]

try:
    elemento = lista[4]
except Exception as e:
    print(f"Ocurrió un error: {e}")
else:
    print(f"El elemento en el índice 4 es: {elemento}")
finally:
    print("La ejecución ha finalizado.")


"""
    DIFICULTAD EXTRA (opcional):
"""


class MiError(Exception):
    pass


def procesar_parametros(a, b):
    if type(a) != int or type(b) != int:
        raise MiError("Los parámetros deben ser enteros.")
    return a / b


try:
    resultado = procesar_parametros(10, "a")
except ZeroDivisionError as e:
    print(f"Ocurrió un error: {e}")
    print(f"El tipo de error es: {type(e)}")
except MiError as e:
    print(f"Ocurrió un error: {e}")
    print(f"El tipo de error es: {type(e)}")
else:
    print(f"El resultado es: {resultado}")
finally:
    print("La ejecución ha finalizado.")
